src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs

```diff
@@ -1578,6 +1584,80 @@ private static bool HasAnonymousFunctionConversion(BoundExpression source, TypeS
+        }
+
+        internal static ConstructibleCollectionTypeKind GetConstructibleCollectionType(CSharpCompilation compilation, TypeSymbol destination, out TypeSymbol? elementType)
+        {
+            Debug.Assert(compilation is { });
+
+            if (destination is ArrayTypeSymbol arrayType)
+            {
+                if (arrayType.IsSZArray)
+                {
+                    elementType = arrayType.ElementType;
+                    return ConstructibleCollectionTypeKind.Array;
+                }
+            }
+            else if (isSpanType(compilation, destination, WellKnownType.System_Span_T, out elementType))
+            {
+                return ConstructibleCollectionTypeKind.Span;
+            }
+            else if (isSpanType(compilation, destination, WellKnownType.System_ReadOnlySpan_T, out elementType))
+            {
+                return ConstructibleCollectionTypeKind.ReadOnlySpan;
+            }
+            else if (supportsCollectionInitializer(compilation, destination))
+            {
+                elementType = null;
+                return ConstructibleCollectionTypeKind.CollectionInitializer;
+            }
+
+            elementType = null;
+            return ConstructibleCollectionTypeKind.None;
+
+            static bool isSpanType(CSharpCompilation compilation, TypeSymbol targetType, WellKnownType spanType, [NotNullWhen(true)] out TypeSymbol? elementType)
+            {
+                if (targetType is NamedTypeSymbol { Arity: 1 } namedType
+                    && TypeSymbol.Equals(namedType.OriginalDefinition, compilation.GetWellKnownType(spanType), TypeCompareKind.AllIgnoreOptions))
+                {
+                    elementType = namedType.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0].Type;
+                    return true;
+                }
+                elementType = null;
+                return false;
+            }
+
+            static bool supportsCollectionInitializer(CSharpCompilation compilation, TypeSymbol targetType)
+            {
+                ImmutableArray<NamedTypeSymbol> allInterfaces;
+                switch (targetType.TypeKind)
+                {
+                    case TypeKind.Class:
+                    case TypeKind.Struct:
+                        allInterfaces = targetType.AllInterfacesNoUseSiteDiagnostics;
+                        break;
+                    case TypeKind.TypeParameter:
+                        allInterfaces = ((TypeParameterSymbol)targetType).AllEffectiveInterfacesNoUseSiteDiagnostics;
+                        break;
+                    default:
+                        return false;
+                }
+
+                // This implementation differs from Binder.CollectionInitializerTypeImplementsIEnumerable().
+                // That method checks for an implicit conversion from IEnumerable to the collection type,
+                // but that would allow: Nullable<StructCollection> s = [];
```

Comment:
Should we adjust the behavior of that method and use it here? It seems like `new StructCollection? { 1, 2, 3 }` is always broken. Currently it seems like this errors because no Add method is available. But it definitely wouldn't make sense for this collection initializer to work, using the `IEnumerable`-ness of the underlying type, and the `Add` method of the nullable type.

Reply:
Agreed. I don't like these being different.

Reply:
Added a PROTOTYPE comment.